File: spec/digit_parser_spec.rb
require 'digit_parser'

describe DigitParser do

  describe "#parse" do

    it "parses digits" do
      parser = DigitParser.new

      expected = {
        " _ " +
        "| |" +
        "|_|"   => '0',
        "   "+
        "  |"+
        "  |"   => '1',
        " _ " +
        " _|" +
        "|_ "   => '2',
        " _ " +
        " _|" +
        " _|"    => '3',
        "   " +
        "|_|" +
        "  |"   => '4',
        " _ " +
        "|_ " +
        " _|"   => '5',
        " _ " +
        "|_ " +
        "|_|"   => '6',
        " _ " +
        "  |" +
        "  |"   => '7',
        " _ " +
        "|_|" +
        "|_|"   => '8',
        " _ " +
        "|_|" +
        " _|"   => '9'
      }
      
      # NOTE: This is a technique I use when testing large numbers of inputs through a functon.
      # The test doesn't abort on the first match failure. The rspec matcher outputs a nice diff
      # showing the failed matches. For a small number of cases like this, it may be clearer to
      # just call parse() for each string. 
      actual = expected.keys.each_with_object({}) do |digit, h|
        h[digit] = parser.parse(digit)
      end
      expect(actual).to eq(expected)
    end
    
    it "returns ? if it doesn't recognize the digit" do
      expect(DigitParser.new.parse('xx')).to eq('?')
    end
  end
end

File: javascript/spec/digit_parser_spec.js
describe("DigitParser", function() {

  describe("#parse", function() {

    it("parses digits", function() {
      var parser = new DigitParser();

      tests = [
        { when : " _ " +
                 "| |" +
                 "|_|",  expect : '0' },
        { when : "   "+
                 "  |"+
                 "  |",  expect : '1' },
        { when : " _ " +
                 " _|" +
                 "|_ ",  expect : '2' },
        { when : " _ " +
                 " _|" +
                 " _|",  expect : '3' },
        { when : "   " +
                 "|_|" +
                 "  |",  expect : '4' },
        { when : " _ " +
                 "|_ " +
                 " _|",  expect : '5' },
        { when : " _ " +
                 "|_ " +
                 "|_|",  expect : '6' },
        { when : " _ " +
                 "  |" +
                 "  |",  expect : '7' },
        { when : " _ " +
                 "|_|" +
                 "|_|",  expect : '8' },
        { when : " _ " +
                 "|_|" +
                 " _|",  expect : '9' }
      ];

      tests.forEach(function(test) {
        expect(parser.parse(test.when)).toEqual(test.expect)
      });
    });

    it("returns ? if it doesn't recognize the digit", function() {
      var parser = new DigitParser();
      expect(parser.parse('xx')).toEqual('?');
    });

  });

});
